pkg/daemon/spawn.go

```diff
@@ -58,7 +59,7 @@ func Spawn(cfg *SpawnConfig) error {
 	binPath = abs("BinPath", binPath)
 	dbPath := abs("DbPath", cfg.DbPath)
 	sockPath := abs("SockPath", cfg.SockPath)
-	logPathPrefix := abs("LogPathPrefix", cfg.LogPathPrefix)
+	logPath := abs("LogPathPrefix", cfg.LogPathPrefix+strconv.Itoa(os.Getpid()))
```

Comment:
Hmm, IIUC this PR moves the opening of the log file from the daemon process to its parent. Since the filename is now determined in the parent, it no longer reflects the PID of the daemon process itself, but that of the parent.

Why is this necessary, and is it related to the main purpose of this PR (ensure the daemon doesn't have fds open on the tty)?

Reply:
Yes, this does what you described. The reason is we don't want to risk losing any messages written to stdout/err before the daemon setup code can open a log file and redirect its os.Stdout/err to that log file. This is also marginally simpler than doing it in the daemon setup code. I'm also convinced that sooner, rather than later, the daemon will be eliminated. Too, this allows eliminating the `-logprefix` flag which shouldn't, and isn't mean to be, used by a person. So the fix employed by this P.R. makes more sense.

That the PID on the file doesn't match that of the daemon isn't really important. Note that the daemon explicitly records its PID in the first log message it writes. Regardless of which process names the file the PID could still match a prior execution and thus a log file that already exists.

Reply:
> The reason is we don't want to risk losing any messages written to stdout/err before the daemon setup code can open a log file and redirect its os.Stdout/err to that log file.

That is not correct. The stdout and stderr of the daemon are not redirected to the log file specified by `-log` or `-logprefix`; those flags affect the destination of the "loggers" defined in https://github.com/elves/elvish/tree/master/pkg/logutil. There is actually a TODO saying that the stdout and stderr of the daemon probably should be redirected, though :): https://github.com/elves/elvish/blob/3615988408b7105c92096a784d915ca2568a83cb/pkg/daemon/spawn.go#L75

> That the PID on the file doesn't match that of the daemon isn't really important.

I agree and disagree. I agree that it's not important to have the PID as part of the filename since the first message is reporting the PID of the daemon. However, having another mysterious number that looks like a PID but is **not** the PID of the daemon can be very confusing.

It's better to just use a serial number; so instead of `daemon.log-23178` or `daemon.log-12865` you get `daemon.log-1`, `daemon.log-2`, etc.. I've actually implemented a function to support this idea, but so far the function is unused:

https://github.com/elves/elvish/blob/3615988408b7105c92096a784d915ca2568a83cb/pkg/fsutil/claim.go#L15

For the sake of this PR, I'd prefer that you leave this part out, since it's not necessary for the more important change of ensuring the daemon doesn't have FDs open on the TTY.

Reply:
> That is not correct. The stdout and stderr of the daemon are not redirected to the log file specified by -log or -logprefix

It is true with this change since I modified the code to pass os.Stdout to the logger. See the block at line 111 in *pkg/prog/prog.go*. I overlooked the TODO. That comment should also be removed.

Reply:
Hold on, I did remove the TODO about redirecting stdout/stderr for the daemon.

Reply:
Please address this feedback:

> That the PID on the file doesn't match that of the daemon isn't really important.

I agree and disagree. I agree that it's not important to have the PID as part of the filename since the first message is reporting the PID of the daemon. However, having another mysterious number that looks like a PID but is **not** the PID of the daemon can be very confusing.

It's better to just use a serial number; so instead of `daemon.log-23178` or `daemon.log-12865` you get `daemon.log-1`, `daemon.log-2`, etc.. I've actually implemented a function to support this idea, but so far the function is unused:

https://github.com/elves/elvish/blob/3615988408b7105c92096a784d915ca2568a83cb/pkg/fsutil/claim.go#L15

Reply:
Okay, but you wrote in your previous comment immediately after talking about possibly using `ClaimFile`:

> For the sake of this PR, I'd prefer that you leave this part out, since it's not necessary for the more important change of ensuring the daemon doesn't have FDs open on the TTY.

Also, these are *temporary* files that are removed on each reboot and it should be rare for users to have more than one of these daemon log files. But since `ClaimFile` already exists I have no objection to using it. Updated PR in a few minutes....

Reply:
FYI, I found a bug in `ClaimFile`. It fails to create the file in the directory that it scanned for existing files that match the pattern. Instead it creates the file in the CWD. I've fixed that but didn't bother to update its unit test.

Reply:
> Okay, but you wrote in your previous comment immediately after talking about possibly using ClaimFile:

Oh, I originally thought that this PR contains two independent changes, (1) making sure that the daemon doesn't have FDs on the TTY and (2) redirecting the stdout and stderr of the daemon, so I was asking you to make this PR only do (1) but not (2) - by "that" I meant part (2). But later I realized that those two things are the same thing.